### ISSUE TITLE
Implement a linear model with stochastic gradient descent learning

### ISSUE THREAD

**Issue by oleksiyskononenko** [h2oai/datatable#2358]
A linear model we are going to add to `datatable.models` should support binomial and multinomial classifications, as well as numeric regression. It should be a regularized model with stochastic gradient decent (SGD) learning similar to the implementation provided by [scikit-learn](https://scikit-learn.org/stable/modules/classes.html#module-sklearn.linear_model).

The model should support all the column types that are currently supported by `datatable`, i.e. numeric and string columns, and will include 

**I. The preparation step: feature binning**
- for string columns we will do one hot encoding, it should also be possible to specify category groups for a particular column;
- for numeric columns we will do binning first, the number of bins or the corresponding intervals will be an input to the model. Second, we will do one hot encoding based on the binned data.
These preparation steps will ensure the model has high interpretability. 

**II. Training the model**
Each row of data will be converted on-the-fly to a set of zeros and ones — an input to our model. Then, we will train the model by using the SGD method. Its API should be similar to the datatable FTRL API: regularization parameters, learning rate, precision, etc. but will exclude parameters related to the hashing trick.

**Comment by XiaomoWu** [comment]
First, let me be clear that I'm more than happy to see new features added to `pydatatable`. However, do you mind talking more about why you consider linear models to be necessary for `pydatatable`, which is primarily a data manipulation package instead of an ML package? I'm just curious about which direction will `pydatatable` take in the future: a dedicated 2D data manipulation tool, or a combination of data manipulation and ML tool? At last, thanks again for bringing the amazing `Rdatatable` to python!

**Comment by oleksiyskononenko** [comment]
Dear @XiaomoWu, yes, the main purpose of datatable is to be a dedicated 2D data manipulation tool. However, we also have a limited number of models under the datatable.models, namely, data aggregator, the FTRL model and the k-fold splitter.

For the moment we are not going to make datatable.models as complete as possible in terms of the ML tools. This is because, first, there are already quite a lot of excellent tools available to the community; second, we do not have enough resources and see no reason to bring all of the existing algos under the same umbrella. 

What we are going to do instead, is to have a limited set of models that would benefit most from being implemented in datatable. For example, keeping both the algo and the data in datatable we were able to develop [the FTRL model](https://datatable.readthedocs.io/en/latest/ftrl.html), that showed pretty [good performance](https://www.kaggle.com/c/microsoft-malware-prediction/discussion/75478) in a variety of use cases. Hence, this is the direction we also want to follow by adding a linear model.

Thanks for your interest and support of the datatable package, I hope you will enjoy the new features we have on our roadmap for the next release.

**Comment by XiaomoWu** [comment]
Hi @oleksiyskononenko, thanks a lot for the explanation! Actually I completely missed the FTRL part when I'm studying the doc of `pydatatable`, which now seems to be a great add-on feature. As you said, there have been a lot of excellent tools out in the community, so personally I'm wishing a lot that `pydatatable` could leverage them by allowing custom aggregation function in `j` in the near future as of #1829. Thank you!
